expression/builtin_time.go

```diff
@@ -305,6 +305,34 @@ type builtinTimeDiffSig struct {
 	baseBuiltinFunc
 }
 
+func (b *builtinTimeDiffSig) getStrFsp(strArg string, fsp int) int {
+	if n := strings.IndexByte(strArg, '.'); n >= 0 {
+		lenStrFsp := len(strArg[n+1:])
+		if lenStrFsp <= types.MaxFsp {
+			fsp = int(math.Max(float64(lenStrFsp), float64(fsp)))
+		}
+	}
+	return fsp
+}
+
+func (b *builtinTimeDiffSig) convertArgToTime(sc *variable.StatementContext, arg types.Datum, fsp int) (t types.Time, err error) {
+	// Fix issue #3923, see https://github.com/pingcap/tidb/issues/3923,
+	// TIMEDIFF() returns expr1 − expr2 expressed as a time value. expr1 and expr2 are time or date-and-time expressions, but both must be of the same type.
+	// if expr is a string, MySQL will try to convert it to mysql time first(in TiDB,
```

Comment:
If expr is a string, we first try to convert it to \`Duration\`,
if it failed to convert, we then try to convert it to \`Datetime\`.